src/Ztm.WebApi/TransactionConfirmationWatcher.cs

```diff
@@ -0,0 +1,479 @@
+            CancellationToken cancellationToken)
+        {
+            if (transaction == null)
+            {
+                throw new ArgumentNullException(nameof(transaction));
+            }
+
+            if (confirmation < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(confirmation) ,"Confirmation is less than zero.");
+            }
+
+            if (!Timer.IsValidDuration(unconfirmedWaitingTime))
+            {
+                throw new ArgumentOutOfRangeException(nameof(unconfirmedWaitingTime), "UnconfirmedWaitingTime is invalid duration.");
+            }
+
+            if (callback == null)
+            {
+                throw new ArgumentNullException(nameof(callback));
+            }
+
+            if (successData == null)
+            {
+                throw new ArgumentNullException(nameof(successData));
+            }
+
+            if (timeoutData == null)
+            {
+                throw new ArgumentNullException(nameof(timeoutData));
+            }
+
+            var rule = await this.ruleRepository.AddAsync
+            (
+                transaction,
+                confirmation,
+                unconfirmedWaitingTime,
+                successData,
+                timeoutData,
+                callback,
+                cancellationToken
+            );
+
+            if (!await TryToRecoverOnChainTransactionAsync(rule, cancellationToken))
+            {
+                // The transaction is not on chain yet.
+                await SetupTimerAsync(rule);
+            }
+
+            return rule;
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            var rules = await this.ruleRepository.ListAsync(cancellationToken);
+
+            foreach (var rule in rules.Where(w => !w.Completed))
+            {
+                if (!await TryToRecoverOnChainTransactionAsync(rule, cancellationToken)) // lgtm[cs/linq/missed-where]
+                {
+                    // The transaction is not on chain yet.
+                    await SetupTimerAsync(rule);
+                }
+            }
+        }
+
+        public async Task StopAsync(CancellationToken cancellationToken)
+        {
+            this.timerLock.EnterWriteLock();
+
+            try
+            {
+                foreach (var timerSet in timers)
+                {
+                    foreach (var timer in timerSet.Value.Where(t => t.Value != null && t.Value.Status == TimerStatus.Started))
+                    {
+                        await timer.Value.StopAsync(cancellationToken);
+                    }
+                }
+            }
+            finally
+            {
+                this.timerLock.ExitWriteLock();
+            }
+        }
+
+        // Check if transaction is on chain, then adding a new watches object.
+        async Task<bool> TryToRecoverOnChainTransactionAsync(Rule rule, CancellationToken cancellationToken)
```

Comment:
If we followed my idea we don't need this. Also this one is some kind of hack too.